### ISSUE TITLE
Dimensional reduction

### ISSUE THREAD

**Issue by bloer** [hgrecco/pint#500]
I was really excited to see how pint handled a lot of "gotcha" issues, until I ran into something that seemed super simple that pint throws a curve on.

Maybe there's something obvious I'm missing, but it seems like pint should be able to recognize when you have a ratio of two quantities with the same dimensionality and automatically do the conversions to reduce them.  For example: 
    In [1]: import pint
    In [2]: ureg = pint.UnitRegistry()
    In [3]: density = 3*ureg.g / ureg.L
    In [4]: volume = 23*ureg.milliliter
    In [5]: mass = density * volume
    In [6]: print(mass)
    69.0 gram * milliliter / liter
    In [7]: mass.dimensionality
    Out[7]: <UnitsContainer({'[mass]': 1.0})>

So pint even correctly recognizes the dimensionality, and is just carrying around this constant "milliliter/liter" factor.  Users shouldn't have to take an extra step to do something as simple as multiply density*volume to get mass or speed*time to get distance

**Comment by dalito** [comment]
The dimensionality reduction is only done if requested. You can use `mass.to_base_units()` ro `mass.to_root_units()` (see [documentation](https://pint.readthedocs.io/en/0.7.2/systems.html)).

This avoids (i) unnecessary computation and (ii) any decision about type conversion. When following your example in python 2, the dimensionality reduction requires a change from type int to another type (float or decimal?).

**Comment by bloer** [comment]
About the type conversion question, this decisions is already made when you call `to_base_units()`, so I think the same logic should apply there.  

Consider another very unintuitive consequence: 
    x = (10*ureg.feet / 3*ureg.inches)
Now I have a quantity that is both `unitless` and `dimensionless`, but I still have to call `to_base_units()` before I can do anything sensible with it.  Naively I would expect `to_base_units()` to be a no-op on a Quantity that is `unitless`. 

Since there is some potentially unnecessary computation done if dimensions are always reduced by default, what about adding a switch somewhere that users could toggle?  Something like a `Quantity.AUTO_REDUCE_DIMENSIONS` class member?

**Comment by dalito** [comment]
> what about adding a switch somewhere that users could toggle? Something like a Quantity.AUTO_REDUCE_DIMENSIONS class member?

Supporting automatic dimensional reduction via an option sound like a nice idea to me (especially for interactive use). 

IIRC pint has no options on Quantity class level. Instead options are set on unit-registry level, e.g. `autoconvert_offset_to_baseunit` (in [registry.py](https://github.com/hgrecco/pint/blob/master/pint/registry.py)). I think your idea could also be added on registry level.

Note that I am not a pint maintainer but only an occasional contributor.

**Comment by dopplershift** [comment]
The counter argument for having reduction as the default is this:

```python
x = (10 * ureg.feet / 3 * ureg.inches)

# Later in another part of the code
y = x * (10 * ureg.inches)
```
If you always reduce, you can invoke unnecessary work, which is non-trivial for large datasets. Having an opt-in flag would be ok. It would also be ok (IMO) if `.magnitude` and calls to `array()` caused dimensionless quantities to auto reduce.

**Comment by hgrecco** [comment]
The original reasons for this being the default behavior are those given by @dalito . In addition, there are some "weird quantities" out there in which you do not want to automatically simplify.

Having said this, it is a feature has been suggested/request in the past but nobody has proposed a clean API and a PR. I will be happy to merge such a change with an opt-in flag.

**Comment by hgrecco** [comment]
Any follow up on this?

**Comment by bloer** [comment]
I'm finding the source code to be daunting; I can't figure out in the short look I've had where to put the basic hooks, much less test corner-cases. 
Between that and very low time availability, I don't think a PR is likely to come from me any time soon...

**Comment by aglassdarkly** [comment]
I was playing around with a solution and got dimension reduction working as expected in the original mass example. However, the second example doesn't appear to be dimensionless as stated.
```python
x = (10 * ureg.feet / 3 * ureg.inches) 
str(x)
3.3333333333333335 foot * inch
```
For a dimensionless quantity, shouldn't it be something like this?
```python
x = (10 * ureg.feet) / (3 * ureg.inches)
str(x)
3.3333333333333335 foot / inch
```
I apologize if I am misunderstanding. I am new to this library.

**Comment by bloer** [comment]
Good catch @aglassdarkly .  You are correct that what I wrote doesn't work.

**Comment by aglassdarkly** [comment]
Thanks @bloer. I will clean up my changes and submit a PR soon.
